src/fetchers.rs

```diff
@@ -44,30 +46,59 @@ pub struct Data {
 
 #[derive(Default)]
 pub struct MultiFetcher {
-    fetchers: Vec<Box<dyn Fetcher>>,
+    fetchers: Vec<Arc<dyn Fetcher>>,
 }
 
 impl MultiFetcher {
-    pub fn add(&mut self, fetcher: Box<dyn Fetcher>) {
+    pub fn add(&mut self, fetcher: Arc<dyn Fetcher>) {
         self.fetchers.push(fetcher);
     }
 
-    pub async fn first_available(&self) -> Option<&dyn Fetcher> {
-        for fetcher in &self.fetchers {
-            let available = fetcher.check().await.unwrap_or_else(|err| {
-                debug!(
-                    "Error while checking fetcher {}: {}",
-                    fetcher.source_name(),
-                    err
-                );
-                false
-            });
-
-            if available {
-                return Some(&**fetcher);
+    pub async fn first_available(&self) -> Option<Arc<dyn Fetcher>> {
+        let handles: Vec<_> = self
+            .fetchers
+            .iter()
+            .cloned()
+            .map(|fetcher| {
+                let fetcher_cloned = fetcher.clone();
+
+                (
+                    AutoAbortJoinHandle(tokio::spawn(async move { fetcher.check().await })),
```

Comment:
Perhaps something like this would be more elegant than cloning the fetcher twice?

```suggestion
                    AutoAbortJoinHandle(tokio::spawn(async move {
                        if fetcher.check().await {
                            Some(fetcher)
                        } else {
                            None
                        }
                    })),
```

Reply:
But we still need `fetchers` to be there when reporting the error, which can be caused by `panic` and reported by `JoinError`.

Reply:
Oh, yeah, fair enough